### ISSUE TITLE
Support single word rule composition

### ISSUE THREAD

**Issue by patapizza** [facebook/duckling#110]
Today Duckling doesn't allow to create multiple patterns to match against a single word.
As a result, we can't compose rules/dimensions and have to duplicate the regexes.

This is the case for DE (e.g. [here](https://github.com/facebook/duckling/blob/master/Duckling/Numeral/DE/Rules.hs#L105) and [here](https://github.com/facebook/duckling/blob/master/Duckling/Numeral/DE/Rules.hs#L323)), NL (e.g. [here](https://github.com/facebook/duckling/blob/master/Duckling/Numeral/NL/Rules.hs#L104)) and EL (e.g. [here](https://github.com/facebook/duckling/blob/master/Duckling/Duration/EL/Rules.hs#L121)).

**Comment by chessai** [comment]
Yes, this is a somewhat annoying thing about hacking on duckling.

**Comment by emlautarom1** [comment]
I've been thinking about how we could approach this issue. My idea is to add a rule to each language that splits single words so we can reuse the rules that already exist.

---

Assume that we have `Language X` and it has rules for `hello`, `world` and `intersection`, so it correctly parses `hello` (with `hello rule`), `world` (with `world` rule), and `hello world`, `world hello`, `hello hello`, `world world` (with `intersection` rule).

What I would like to do is add a `single-word decomposition` rule that will handle `helloworld`, `hellohello`, `worldhello` and `worldworld`. The idea is that this rule will **only** split the word into subwords which then will be picked up by the `intersection` rule.

For example, rule `single-word decomposition` converts `helloworld` into `hello world`. Then, rule `intersection` picks `hello world` and works as expected.

The `single-word decompositon` rule would need to know in advance **all possible subwords** so then it can separate them. A rough implementation would be something like:

```haskell
ruleSingleWordDecomposition :: Rule
ruleSingleWordDecomposition = Rule
  { name = "single-word decomposition"
  , pattern =
    [ regex "(\\S+)" -- we match 'single' words
    ]
  , prod = \tokens -> case tokens of
      (Token RegexMatch (GroupMatch (w:_)):_) -> do
        let rgx = "(hello|world)" :: String -- we need to compute in advance a big regex of possible "subwords"
        let ws = Text.pack <$> getAllTextMatches (show w =~ rgx)
        Just $ Token RegexMatch $ GroupMatch ws -- How do we return the "subwords" so other rules can pick them up?
      _ -> Nothing
  }
```

*Note: we need to split the "single word" in Haskell* (`let ws = Text.pack <$> getAllTextMatches (show w =~ rgx)`) *since we can't do it with pure regexes AFAIK*

In this case, our `single-word decomposition` rule knows that "single words" can be composed only from the "subwords" `hello` and `world`.

The main issue that I found with this is **how to use the output of this rule as input for the remaining rules** (they don't pick it up). So, even if we're able to build the required regex and split the "single word" into "subwords" I can't seem to redirect them to other rules.

I'm trying to avoid adding special cases or hacking into the internals of Duckling (`Engine.hs` and the like), so any hint would be appreciated.

**Comment by emlautarom1** [comment]
One quick note, the mentioned examples in the first comment were not permalinks so I can't be sure to what do they refer. I understand that this issue is related to:

- https://github.com/facebook/duckling/issues/481
- https://github.com/facebook/duckling/issues/415
